source/common/common/utility.h

```diff
@@ -636,13 +636,72 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// A TrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
+// child will contain a vector of size 1. A node with no children will
+// contain an empty vector.
+//
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
+template <class Value> class TrieEntry {
+public:
   Value value_{};
-  std::array<std::unique_ptr<TrieEntry>, 256> entries_;
+  const TrieEntry* operator[](uint8_t i) const {
+    if (i >= min_child_ && i < (min_child_ + children_.size())) {
+      return children_[i - min_child_].get();
+    }
+    return nullptr;
+  }
+  TrieEntry* operator[](uint8_t i) { return const_cast<TrieEntry*>(std::as_const(*this)[i]); }
+
+  void set(uint8_t branch, std::unique_ptr<TrieEntry<Value>> entry) {
```

Comment:
For this, please make sure to comment the meaning of `branch`.

Reply:
Renamed all the key parameters to key, then rethought that because "key" being used for both an entry key and a child-branch-index made it more confusing if anything, so all the node-child-branch-index arguments are named `c` and commented accordingly.

Also renamed `TrieEntry` to `TrieNode`, because most of them are not actually entries, so now there's distinct concepts instead of blurry confusion.

(Had originally left it as `TrieEntry` because that's how it was, not because I thought it was a good idea, but if we're cleaning up all the existing mess, might as well do more!)

Reply:
Big +1 on the TrieEntry -> TrieNode rename. That seems cleaner.

But I'm less clear about the `c`. I too like the name `key`, but I'm not quite sure I appreciate the distinction between a "entry key and a child-branch-index". Can you say more about this? Is the idea that the "entry key" is the full string in the `TrieLookupTable` whereas `child-branch-index` is the `[0-256)` index of the child in the node? If so, I can see not using key for both. In that case, I think I would suggest `child_index` (since that matches up well with its use and an index into `children_`.) WDYT?

Reply:
+1 for child_index.